File: Java/Hired Programming Practice/src/StacksAndQueues/Q3_3.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package StacksAndQueues;

import java.util.ArrayList;

/**
 *
 * @author marcsantiago
 */
//Implement a data structure called setofstacks that mimics this.  SetOfStacks should
//be composed of several stacks and should create a new stack once the previous
//one exceeds capacity.  Push and pop should act as if there is just one stack
public class Q3_3 {
    
    ArrayList<Stack2> stacks = new ArrayList<>();    
    
    void push(int d){
        Stack2 last = getLastStack();
        if(last != null && !last.isFull()){
            last.push(d);
        }
        else{
            Stack2 stack = new Stack2();
            stack.push(d);
            stacks.add(stack);
        }
    }
        
    int pop(){
        Stack2 last = getLastStack();
        int v = last.pop();
        if(last.size == 0){
            stacks.remove(stacks.size() - 1);
        }
        return v;
        
    }
    
    int popAt(int index){
        if(index + 1 > numberOfStacks()){
            return -1;
        }
        Stack2 myStack = stacks.get(index);
        int v = myStack.pop();
        if(myStack.size == 0){
            stacks.remove(stacks.get(index));
        }
        return v;
    }
 
    int numberOfStacks(){
        return stacks.size();
    }
    
    int peek(){
        Stack2 last = getLastStack();
        return last.top.data;
    }
    
    Stack2 getLastStack(){
       if(stacks.isEmpty()){
           return null;
       }
       else{
          int last = stacks.size() - 1;
          return stacks.get(last); 
       } 
    }
    
}

class Node3 {
    int data;
    Node3 next = null;
    public Node3(int d){
        data = d;
    }
}

class Stack2{
    Node3 top;
    int capacity = 10;
    int count = 0;
    int size = 10;
    
    int pop(){
        if(top != null){
            int item = top.data;
            top = top.next;
            size--;
            return item;
        }
        return -1;
    }
    
    void push(int item){
        Node3 t = new Node3(item);
        t.next = top;
        top = t;
        count++;
    }
    
    boolean isFull(){
        return count == capacity;
    }
    
    boolean isEmpty(){
        return top == null;
    }
    
    int peek(){
        return top.data;
    }
}

File: Python/Stack/Q3_3.py
# Implement a data structure called SetOfStacks that mimics this.  SetOfStacks should
# be composed of several stacks and should create a new stack once the previous
#one exceeds capacity.  Push and pop should act as if there is just one stack

class Node(object):
    def __init__(self):
        self.data = None
        self.next = None


class SetOfStacks(object):
    # not really necessary, nodes just make it messy, pythons list can do the work,
    # but the problem said to make a stack structure
    top = Node()
    capacity = 10
    count = 0
    stacks = list()
    s = list()

    def pop(self):
        if len(self.s) is None or len(self.stacks) is None:
            return None
        if len(self.s):
            v = self.s.pop()
            return v
        else:
            v = self.stacks[self.get_last_stack()].pop()
            return v

    def pop_at(self, index):
        if self.get_last_stack() is None or index > self.get_last_stack():
            return None
        if self.s != [] and index == 0 and self.get_last_stack() is None:
            return self.s.pop()
        else:
            return self.stacks[index].pop()


    def push(self, item):
        n = Node()
        n.data = item

        if not self.is_full():
            self.top = n
            self.s.append(self.top.data)
        else:
            self.stacks.append(self.s)
            self.reset_stack()
            self.s = []
        self.count += 1


    def peek(self):
        if len(self.s) is None or len(self.stacks) is None:
            return None
        if len(self.s):
            return self.s[-1]
        else:
            v = self.stacks[self.get_last_stack()]
            return v[-1]

    def is_empty(self):
        return self.top.data is None

    def is_full(self):
        return self.capacity == self.count

    def reset_stack(self):
        self.count = 0

    def get_last_stack(self):
        if (len(self.stacks)) == 0:
            return None
        return len(self.stacks) - 1


